src/combinatorics/quiver.lean

```diff
@@ -3,130 +3,139 @@ Copyright (c) 2021 David Wärn. All rights reserved.
+{ S := Type u,
+  coe := λ H, wide_subquiver.to_Type V H, }
 
 /-- A wide subquiver viewed as a quiver on its own. -/
-def wide_subquiver.quiver {V} {G : quiver V} (H : wide_subquiver G) : quiver V :=
+instance wide_subquiver.quiver {V} [quiver V] (H : wide_subquiver V) : quiver H :=
 ⟨λ a b, H a b⟩
 
 namespace quiver
 
-/-- The quiver with no arrows. -/
-protected def empty (V) : quiver.{v} V := ⟨λ a b, pempty⟩
-
-instance {V} : inhabited (quiver.{v} V) := ⟨quiver.empty V⟩
-instance {V} (G : quiver V) : has_bot (wide_subquiver G) := ⟨λ a b, ∅⟩
-instance {V} (G : quiver V) : has_top (wide_subquiver G) := ⟨λ a b, set.univ⟩
-instance {V} {G : quiver V} : inhabited (wide_subquiver G) := ⟨⊤⟩
-
-/-- `G.sum H` takes the disjoint union of the arrows of `G` and `H`. -/
-protected def sum {V} (G H : quiver V) : quiver V :=
-⟨λ a b, G.arrow a b ⊕ H.arrow a b⟩
-
-/-- `G.opposite` reverses the direction of all arrows of `G`. -/
-protected def opposite {V} (G : quiver V) : quiver V :=
-⟨flip G.arrow⟩
+/-- A type synonym for a quiver with no arrows. -/
+@[nolint has_inhabited_instance]
+def empty (V) : Type u := V
 
-/-- `G.symmetrify` adds to `G` the reversal of all arrows of `G`. -/
-def symmetrify {V} (G : quiver V) : quiver V :=
-G.sum G.opposite
+instance empty_quiver (V : Type u) : quiver.{u} (empty V) := ⟨λ a b, pempty⟩
 
-@[simp] lemma empty_arrow {V} (a b : V) : (quiver.empty V).arrow a b = pempty := rfl
+@[simp] lemma empty_arrow {V : Type u} (a b : empty V) : (a ⟶ b) = pempty := rfl
 
-@[simp] lemma sum_arrow {V} (G H : quiver V) (a b : V) :
-  (G.sum H).arrow a b = (G.arrow a b ⊕ H.arrow a b) := rfl
+instance {V} [quiver V] : has_bot (wide_subquiver V) := ⟨λ a b, ∅⟩
+instance {V} [quiver V] : has_top (wide_subquiver V) := ⟨λ a b, set.univ⟩
+instance {V} [quiver V] : inhabited (wide_subquiver V) := ⟨⊤⟩
 
-@[simp] lemma opposite_arrow {V} (G : quiver V) (a b : V) :
-  G.opposite.arrow a b = G.arrow b a := rfl
+/-- `Vᵒᵖ` reverses the direction of all arrows of `V`. -/
+instance op_quiver {V} [quiver V] : quiver Vᵒᵖ :=
+⟨λ a b, (unop b) ⟶ (unop a)⟩
 
-@[simp] lemma symmetrify_arrow {V} (G : quiver V) (a b : V) :
-  G.symmetrify.arrow a b = (G.arrow a b ⊕ G.arrow b a) := rfl
+/-- A type synonym for the symmetrized quiver (with an arrow both ways for each original arrow). -/
+@[nolint has_inhabited_instance]
+def symmetrify (V) : Type u := V
 
-@[simp] lemma opposite_opposite {V} (G : quiver V) : G.opposite.opposite = G :=
-by { cases G, refl }
+instance symmetrify_quiver (V : Type u) [quiver V] : quiver (symmetrify V) :=
+⟨λ a b : V, (a ⟶ b) ⊕ (b ⟶ a)⟩
 
-/-- `G.total` is the type of _all_ arrows of `G`. -/
-@[ext] protected structure total {V} (G : quiver.{v u} V) : Sort (max (u+1) v) :=
+/-- `total V` is the type of _all_ arrows of `V`. -/
+@[ext, nolint has_inhabited_instance]
+structure total (V : Type u) [quiver.{v} V] : Type (max u v) :=
 (source : V)
 (target : V)
-(arrow : G.arrow source target)
```

Comment:
If the intention is to eventually use the `quiver` class instead of `has_hom`, should be rename these fields as `left`, `right` and `hom` to match up the current names around `arrow C`? (This could be left to a later PR as well, as presumably one would want to refactor `arrow C` to use `total`.)

Reply:
I'll make this change in #7151, and add a comment there about eventually unifying with `arrow`. This may be tricky, as at the moment we have `def arrow := comma.{v v v} (𝟭 T) (𝟭 T)`.

Reply:
Marking resolved here.